doc/read-the-docs-site/doc/marconi-as-a-library/tutorials/write-first-indexer.rst

```diff
@@ -1,6 +1,122 @@
+.. _my-first-indexer:
+
 Write your first indexer
 ========================
 
 In this tutorial, we'll show you how to write the simplest possible indexer, and how to expose the query interface of your indexer from a JSON-RPC server.
 
-Coming eventually!
+In order to understand the tutorial, you need to understand:
+
+* Cardano blockchain (specifically `cardano-node`)
+* Haskell (mainly the cardano-api, lenses, sqlite-simple packages)
+* how chain-indexing works in general
+
+The main building blocks of an indexer are:
+
+1) the event, which is the information you want to extract from a block and store it
+2) the instanciation of the backend
+3) the instanciation of the worker
+4) the query interface
+
+For reference, the import list is:
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART import
+   :end-before: BLOCKEND import
+
+Now, let's start with the definition of the event.
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART event
+   :end-before: BLOCKEND event
+
+Given this event, we can instanciate the backend.
+Let's start with the simplest one: the ``ListIndexer``.
+The ``ListIndexer`` backend simply stores each event inside a ``Data.List``, and it is simply instanciated with ``Core.mkListIndexer``.
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART ListIndexer
+   :end-before: BLOCKEND ListIndexer
+
+Once we have instanciated the backend with event type (``BlockInfoEvent``), we can wrap the indexer inside a ``Core.WorkerIndexer`` which is used to spin up a concurrent thread when reading blocks from the local Cardano node using the chain-sync protocol.
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART ListIndexer worker
+   :end-before: BLOCKEND ListIndexer worker
+
+We now have a complete indexer which can be run with ``Core.runIndexer``.
+Next, let's start by defining a runtime environment type which will contain all necessary inputs to run the indexer:
```

Comment:
Although, `runIndexer` takes an `indexer` not a `worker`. The main purpose of the wrapping things in a worker is to set up a coordinator. I think we have two options here:

1. Pass workers and use the indexer directly
2. Briefly Introduce `coordinators` as a way to run several workers.

Reply:
Good point. I'll introduce the coordinator in a separate tutorial

Reply:
Actually, the Worker that good support for updating an `MVar` whenever it processes inputs, and that `MVar` is used by the query thread in order to query to indexer. A lot of manual code needs to be written in order to work with the indexer directly. Unless I misunderstood something?

Reply:
That's true. So we need to include the coordinator at the moment. Maybe we should consider an alternative in Marconi and include it in the tutorial in the future.